File: GraylogTransferLayer/server/src/main/kotlin/com/playdarium/graylogtransferserver/exchange/request/ConstantDataReq.kt
package com.playdarium.graylogtransferserver.exchange.request

import com.playdarium.byteformatter.ByteReader
import com.playdarium.graylogtransferserver.exchange.Header
import com.playdarium.graylogtransferserver.exchange.Request
import com.playdarium.graylogtransferserver.exchange.RequestReader
import com.playdarium.graylogtransferserver.model.ConstantData

class ConstantDataReq(val constantData: ConstantData) : Request {

    override val header: Byte = Header.ConstantData.value

    class Reader : RequestReader {
        override fun readRequest(remoteHost: String, reader: ByteReader): Request {
            val constantData = ConstantData()

            constantData.source = reader.readString()
            constantData.os = reader.readString()
            constantData.osVersion = reader.readString()
            constantData.device = reader.readString()
            constantData.deviceUid = reader.readString()
            constantData.resolution = reader.readString()
            constantData.appVersion = reader.readString()
            constantData.appId = reader.readString()
            constantData.appName = reader.readString()
            constantData.cpu = reader.readString()
            constantData.opengl = reader.readString()
            constantData.memorySize = reader.readShort()
            constantData.gpuMemorySize = reader.readShort()
            constantData.orientation = reader.readByte()

            return ConstantDataReq(constantData)
        }
    }
}

File: Assets/PdLogger/Net/Requests/ConstantDataReq.cs
using ByteFormatter.Runtime;
using PdLogger.Core;
using PdNetwork.Exchange;

namespace PdLogger.Net.Requests
{
    public readonly struct ConstantDataReq : IRequest
    {
        private readonly ConstantData _constantData;

        public ConstantDataReq(ConstantData constantData)
        {
            _constantData = constantData;
        }
        
        public byte GetHeader() => 3;

        public void WriteBody(ByteWriter writer)
        {
            writer.Write(_constantData.Source);
            writer.Write(_constantData.Os);
            writer.Write(_constantData.OsVersion);
            writer.Write(_constantData.Device);
            writer.Write(_constantData.DeviceUid);
            writer.Write(_constantData.Resolution);
            writer.Write(_constantData.AppVersion);
            writer.Write(_constantData.AppId);
            writer.Write(_constantData.AppName);
            writer.Write(_constantData.Cpu);
            writer.Write(_constantData.Opengl);
            writer.Write(_constantData.MemorySize);
            writer.Write(_constantData.GpuMemorySize);
            writer.Write(_constantData.ScreenOrientation);
        }
    }
}